server/src/server.ts

```diff
@@ -0,0 +1,314 @@
+});
+
+/*
+ * Pickup on configuration changes of the client
+ */
+connection.onDidChangeConfiguration((change): void => {
+  // We'll use the path to D2 so we don't get divergent
+  // functionality using two different d2 binaries.
+  d2ExePath = change.settings.D2.execPath;
+  connection.console.info(`Language Server D2 path: ${d2ExePath}`);
+});
+
+/**
+ * The content of a text document has changed. This event is emitted
+ * when the text document first opened or when its content has changed.
+ */
+documents.onDidChangeContent((change: TextDocumentChangeEvent<TextDocument>): void => {
+  // Save the current working directory, which is where
+  // the D2 document is
+  cwd = path.dirname(change.document.uri);
+  const cwDir = URI.parse(cwd).fsPath;
+
+  // Run D2 in the stdin/stdout mode
+  //
+  const args: string[] = ["-"];
+
+  /**
+   * Run D2 in the special "D2_LSP_MODE" mode.
+   */
+  const proc = spawnSync(d2ExePath, args, {
+    input: change.document.getText(),
+    encoding: "utf-8",
+    maxBuffer: 1024 * 1024 * 2,
+    cwd: cwDir,
+    env: { ...process.env, D2_LSP_MODE: "1" },
+  });
+
+  // Pass error back to the client
+  if (proc.pid === 0) {
+    connection.window.showErrorMessage(
+      `Could not find D2 executable! Path: '${d2ExePath}'`
+    );
+    return;
+  }
+
+  // Reset Document Data and Parse out the json from the D2 program
+  astData = new AstReader(proc.stdout);
+
+  /**
+   * Handle any errors.
+   */
+  const eret = astData.Errors;
+
+  // Clear Errors
+  connection.sendDiagnostics({ uri: change.document.uri, diagnostics: [] });
+  if (eret) {
+    eret.uri = change.document.uri;
+    connection.sendDiagnostics(eret);
+  }
+});
+
+/**
+ * If a trigger character is pressed, this is called
+ */
+connection.onCompletion((params: CompletionParams): CompletionList => {
+  switch (params.context?.triggerCharacter) {
+    case "@":
+      return CompletionHelper.doImport(cwd, params.textDocument.uri);
+
+    case ".":
+      return CompletionHelper.doDot(astData, params.position);
+
+    case " ": {
+      const doc = documents.get(params.textDocument.uri);
+      const x = doc?.offsetAt(params.position);
+      const c = doc?.getText().at((x || 2) - 2);
+
+      if (c === ":") {
+        return CompletionHelper.doAttribute(astData, params.position);
+      }
+      break;
+    }
+
+    case undefined:
+      return CompletionHelper.doOpenSpace();
+  }
+
+  return CompletionList.create([], false);
+});
+
+/**
+ * May not need this
+ */
+connection.onCompletionResolve((item: CompletionItem): CompletionItem => {
+  return item;
+});
+
+/**
+ * Request for all document links
+ */
+connection.onDocumentLinks((): DocumentLink[] => {
+  const retLinks: DocumentLink[] = [];
+
+  for (const link of astData.LinksAndImports) {
+    const docLink = DocumentLink.create(link.Range);
+
+    let docPath = link.str.replace(/['|"]/g, "");
+
+    // If '.d2' is there, get rid of it.
+    docPath = path.basename(docPath, ".d2");
```

Comment:
this function should check what the extension is, if it is not d2, this breaks the link to it (and tries to go to file.txt.d2 instead.